Firestore/Source/Core/FSTFirestoreClient.mm

```diff
@@ -81,68 +82,72 @@ - (instancetype)initWithDatabaseInfo:(const DatabaseInfo &)databaseInfo
+   * the user (via public API) or the network (incoming gRPC messages), we should always dispatch
+   * onto this queue. This ensures our internal data structures are never accessed from multiple
+   * threads simultaneously.
+   */
+  std::unique_ptr<AsyncQueue> _workerQueue;
+
   std::unique_ptr<Executor> _userExecutor;
 }
 
 - (Executor *)userExecutor {
   return _userExecutor.get();
 }
 
+- (AsyncQueue *)workerQueue {
+  return _workerQueue.get();
+}
+
 + (instancetype)clientWithDatabaseInfo:(const DatabaseInfo &)databaseInfo
                         usePersistence:(BOOL)usePersistence
                    credentialsProvider:
                        (CredentialsProvider *)credentialsProvider  // no passing ownership
                           userExecutor:(std::unique_ptr<Executor>)userExecutor
-                   workerDispatchQueue:(FSTDispatchQueue *)workerDispatchQueue {
+                           workerQueue:(std::unique_ptr<AsyncQueue>)workerQueue {
   return [[FSTFirestoreClient alloc] initWithDatabaseInfo:databaseInfo
                                            usePersistence:usePersistence
                                       credentialsProvider:credentialsProvider
                                              userExecutor:std::move(userExecutor)
-                                      workerDispatchQueue:workerDispatchQueue];
+                                              workerQueue:std::move(workerQueue)];
 }
 
 - (instancetype)initWithDatabaseInfo:(const DatabaseInfo &)databaseInfo
                       usePersistence:(BOOL)usePersistence
                  credentialsProvider:
                      (CredentialsProvider *)credentialsProvider  // no passing ownership
                         userExecutor:(std::unique_ptr<Executor>)userExecutor
-                 workerDispatchQueue:(FSTDispatchQueue *)workerDispatchQueue {
+                         workerQueue:(std::unique_ptr<AsyncQueue>)workerQueue {
   if (self = [super init]) {
     _databaseInfo = databaseInfo;
     _credentialsProvider = credentialsProvider;
     _userExecutor = std::move(userExecutor);
-    _workerDispatchQueue = workerDispatchQueue;
+    _workerQueue = std::move(workerQueue);
 
     auto userPromise = std::make_shared<std::promise<User>>();
     bool initialized = false;
 
     __weak __typeof__(self) weakSelf = self;
-    auto credentialChangeListener = [initialized, userPromise, weakSelf,
-                                     workerDispatchQueue](User user) mutable {
+    auto credentialChangeListener = [initialized, userPromise, weakSelf](User user) mutable {
       __typeof__(self) strongSelf = weakSelf;
       if (!strongSelf) return;
 
       if (!initialized) {
         initialized = true;
         userPromise->set_value(user);
       } else {
-        [workerDispatchQueue dispatchAsync:^{
+        strongSelf->_workerQueue->Enqueue([weakSelf, user] {
+          __typeof__(self) strongSelf = weakSelf;
```

Comment:
Treating `self` and `this` as equivalent isn't really problematic--they're essentially functionally equivalent.

The difficulty in translating Objective-C to C++ is that when it comes to any Objective-C reference (`self` or otherwise), any assignment to a local or parameter could extend the lifetime of the object. (However, in the interest of not giving the wrong impression the `self` passed implicitly to a method on invocation is `unsafe_unretained`: Objective-C assumes the caller of a method on an object retains the reference).

One `weakSelf` is enough. There's nothing magical about which weak reference we're using: all the strong and weak references to a given object share the same control block, no matter how they're created.